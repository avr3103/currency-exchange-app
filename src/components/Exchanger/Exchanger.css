.exchanger {
  width: 500px;
  background-color: var(--bg-color);
  color: var(--font-color);
  min-height: 320px;
  position: relative;
}

@media (max-width: 600px) {
  .exchanger {
    width: 100%;
    margin: 0 auto;
  }
}

.header-row {
  display: flex;
  align-items: center;
  padding: 20px;
}

.center {
  flex-basis: 50%;
  display: flex;
  justify-content: center;
}

.left {
  flex-basis: 25%;
  display: flex;
  justify-content: flex-start;
  visibility: hidden;
}

.right {
  flex-basis: 25%;
  display: flex;
  justify-content: flex-end;
}

.exchange-btn {
  background: var(--primary-color);
  border: 1px solid var(--primary-color);
  color: var(--font-color);
  padding: 10px;
  cursor: pointer;
}

.exchange-btn:hover {
  background: var(--hover-color);
}

.switch-wallet {
  background: url("../../assets/icons/switch-btn-icon.svg") 50% 50% no-repeat var(--primary-color);
  border-radius: 2.5rem;
  top: 9.6rem;
  position: absolute;
  margin: 0 auto;
  width: 2.5rem;
  height: 2.5rem;
  left: 0;
  right: 0;
  cursor: pointer;
  color: var(--font-color);
  border: none;
}

.delimiter {
  border-bottom: 1px solid var(--font-color);
  margin-bottom: 25px;
}
