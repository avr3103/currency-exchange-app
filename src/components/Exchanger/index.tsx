import {
  FC,
  useEffect,
  ChangeEvent,
  useState,
  useRef
} from 'react';
import { Wallet } from '../Wallet';
import { Button } from '../Button';
import { Loader } from '../Loader';
import { useRatesFetcher } from '../../hooks/useRatesFetcher';
import { useWalletsState } from '../../hooks/useWalletsState';
import {
  getCurrencySymbol,
  calculateRate,
  checkOperation,
  toFixed
} from '../../utils';

import './Exchanger.css';

const Exchanger: FC = (): JSX.Element => {
  const [rates, loading] = useRatesFetcher();
  const [store, dispatch] = useWalletsState();
  const [isExchangeAvailable, setExchangeButton] = useState(false);
  const [valueFrom, setValueFrom] = useState<number | string>('');
  const [valueTo, setValueTo] = useState<number| string>('')

  const {
    wallet: {
      currencies
    },
    currencyFrom,
    currencyTo,
  } = store;

  const inputRef = useRef<HTMLInputElement>(null);

  const setFocus = () => {
    if(inputRef && inputRef.current) {
      inputRef.current.focus();
    }
  };

  useEffect(()=> {
    setFocus();
  },[loading]);

  if (loading) return <Loader/>

  const currencySymbolFrom = getCurrencySymbol(currencyFrom);
  const currencySymbolTo = getCurrencySymbol(currencyTo);
  const [currentRateFrom, currentRateTo] = calculateRate(rates[currencyFrom], rates[currencyTo]);

  const clearInputs = () => {
    if (valueFrom || valueTo) {
      setValueFrom('')
      setValueTo('');
      setExchangeButton(false);
    }
  }

  const setInputValue = (val: string, rate: number, amount: number, wallet: string) => {
    const isCorrectInput = /^[\d]*\.?[\d]{0,2}$/.test(val);
    const res = Number(val) * rate;
    const isOperationCorrect = checkOperation(+val, res, amount);

    setExchangeButton(isOperationCorrect);

    if (isCorrectInput && wallet === 'from') {
      setValueFrom(val ? toFixed(+val) : '');

      if (isOperationCorrect) setValueTo(toFixed(res));
    } else if (isCorrectInput && wallet === 'to'){
      setValueTo(val ? toFixed(+val) : '');

      if (isOperationCorrect) setValueFrom(toFixed(res));
    }

    if (!valueFrom && valueTo) {
      setValueTo('');
    }
  }

  const handleValueChange = (e: ChangeEvent<HTMLInputElement>) => {
    if (e.target.name === 'walletFrom') {
      setInputValue(e.target.value, currentRateFrom, currencies[currencyFrom], 'from')
    } else {
      setInputValue(e.target.value, currentRateTo, currencies[currencyTo], 'to')
    }
  }

  const setCurrentValue = (e: ChangeEvent<HTMLSelectElement>) => {
    const ACTION = e.target.name === 'walletFrom' ? 'CHANGE_FROM': 'CHANGE_TO';
    dispatch({type: ACTION, payload: e.target.value});
    clearInputs();
    setFocus();
  }

  const switchWallets = () => {
    dispatch({type:'SWITCH'});
    clearInputs();
    setFocus();
  }

  const handleExchange = () => {
    const newWallet = {
      currencies: {
        ...currencies,
        [currencyFrom]: toFixed(currencies[currencyFrom] - (+valueFrom)),
        [currencyTo]: toFixed(currencies[currencyTo] + (+valueTo))
      }
    }

    dispatch({type: 'EXCHANGE', payload: newWallet})
    clearInputs();
    setFocus();
  }

  return (
    <div className='exchanger'>
      <div className='header-row'>
        <div className='left' />
        <div className='center'>
          {`${currencySymbolFrom}1 = ${currencySymbolTo}${currentRateFrom.toFixed(2)}`}
        </div>
        <div className='right'>
          <Button
            cname='exchange-btn'
            isActive={isExchangeAvailable}
            text='Exchange'
            onClick={handleExchange}
          />
        </div>
      </div>
      <Wallet
        inputRef={inputRef}
        walletName='walletFrom'
        inputValue={valueFrom}
        currency={currencyFrom}
        currencySymbol={currencySymbolFrom}
        amount={currencies[currencyFrom]}
        onChangeCurrency={setCurrentValue}
        onChangeAmount={handleValueChange}
      />
      <div className='delimiter' />
      <Button cname='switch-wallet' isActive={true} onClick={switchWallets} />
      <Wallet
        walletName='walletTo'
        inputValue={valueTo}
        currency={currencyTo}
        currencySymbol={currencySymbolTo}
        amount={currencies[currencyTo]}
        reteTo={currentRateTo.toFixed(2)}
        symbolFrom={currencySymbolFrom}
        onChangeCurrency={setCurrentValue}
        onChangeAmount={handleValueChange}
      />
    </div>
  );
}



export default Exchanger;