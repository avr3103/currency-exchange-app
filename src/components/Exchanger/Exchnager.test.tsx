import React from 'react';
import { render, screen } from '@testing-library/react';
import Exchanger from './index';

describe('Exchanger tests',() => {
  it('renders Exchanger', () => {
    render(<Exchanger/>);
    const { asFragment } = render(<Exchanger/>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('shows loader', () => {
    render(<Exchanger />);

    const element = screen.getByTestId('loader')
    expect(element).toBeInTheDocument()
  })
})