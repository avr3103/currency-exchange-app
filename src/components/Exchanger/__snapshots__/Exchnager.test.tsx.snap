// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchanger tests renders Exchanger 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
    data-testid="loader"
  >
    <div
      class="dual-ring"
    />
  </div>
</DocumentFragment>
`;
